### PR TITLE
The code for Emiru's real streaming Transformer

### PR DESCRIPTION
**Pull request by D-Keqi**

### CODE REVIEW

**Comment by D-Keqi** [comment]
The code is for Emiru's real streaming Transformer. Most implementations are completed by Hiromasa Fujihara and Emiru Tsunoo. Now, it works.

**Comment by b-flo** [comment]
Thanks @D-Keqi 

@sw005320 @eml914 Should I assume it's the streaming approach we want to follow for ESPnet? I'm asking because I'm currently working on streaming RNN/Conformer Transducer for ESPnet1, if so I'll re-base my implementation upon that.

**Comment by sw005320** [comment]
> @sw005320 @eml914 Should I assume it's the streaming approach we want to follow for ESPnet? I'm asking because I'm currently working on streaming RNN/Conformer Transducer for ESPnet1, if so I'll re-base my implementation upon that.

Yes.
Note that this is block-wise streaming and not frame-wise streaming.

**Comment by sw005320** [comment]
Thanks @D-Keqi, could you please fix the CI error?
https://github.com/espnet/espnet/pull/3614/checks?check_run_id=3674510084

**Comment by b-flo** [comment]
> Note that this is block-wise streaming and not frame-wise streaming.

Got it, thanks! I suppose the related paper is: https://arxiv.org/pdf/2006.14941.pdf

I'm fixing quantization and N-gram decoding for Transducer and I'll review the PR.

File: egs2/TEMPLATE/asr1/asr_streaming.sh [hunk 1]
```diff
@@ -0,0 +1,1383 @@
+#!/usr/bin/env bash
```
Comment (sw005320) [inline egs2/TEMPLATE/asr1/asr_streaming.sh:1]: Do we need to include a training part in the streaming process?
@eml914 @hfujihara, what do you think?

I think we only need an inference part and just re-use `asr.sh` to train the streaming model with the stream model config in each recipe.
However, if this split is difficult, we can keep this file.

**Comment by D-Keqi** [comment]
> > Note that this is block-wise streaming and not frame-wise streaming.
> 
> Got it, thanks! I suppose the related paper is: https://arxiv.org/pdf/2006.14941.pdf
> 
> I'm fixing quantization and N-gram decoding for Transducer and I'll review the PR.

Sure!

File: egs2/aishell/asr1/conf/decode_asr_streaming2.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+beam_size: 20
```
Comment (sw005320) [inline egs2/aishell/asr1/conf/decode_asr_streaming2.yaml:1]: what is the difference of streaming.yaml and streaming2.yaml?
If either of them is unnecessary, please remove it.

**Comment by D-Keqi** [comment]
> Thanks @D-Keqi, could you please fix the CI error?
> https://github.com/espnet/espnet/pull/3614/checks?check_run_id=3674510084

Sure!

File: egs2/aishell/asr1/conf/train_asr_streaming_conformer.yaml [hunk 3]
```diff
@@ -0,0 +1,95 @@
+# network architecture
```
Comment (sw005320) [inline egs2/aishell/asr1/conf/train_asr_streaming_conformer.yaml:1]: Could you emphasize which part is important for the streaming configuration?
(e.g., `contextual_block_conformer`, right?)
Please add such comments.

File: egs2/aishell/asr1/conf/train_asr_streaming_transformer2.yaml [hunk 4]
```diff
@@ -0,0 +1,72 @@
+# network architecture
```
Comment (sw005320) [inline egs2/aishell/asr1/conf/train_asr_streaming_transformer2.yaml:1]: ditto.
What is the difference between transformer.yaml  and transformer2.yaml ?

File: egs2/aishell/asr1/run_streaming.sh [hunk 5]
```diff
@@ -20,7 +20,7 @@ use_wordlm=false
 # (train_set will be "${train_set}_sp" if speed_perturb_factors is specified)
 speed_perturb_factors="0.9 1.0 1.1"
 
-./asr.sh                                               \
+./asr_streaming.sh                                               \
```
Comment (sw005320) [inline egs2/aishell/asr1/run_streaming.sh:23]: If my idea of including the inference part in  `asr_streaming.sh` would work, then we can change here to

```
./asr.sh
...
...

./asr_streaming.sh
...
...
```

File: egs2/laborotv/asr1/run_streaming.sh [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+asr_config=conf/train_asr_streaming_transformer.yaml
+inference_config=conf/decode_asr_streaming.yaml
+lm_config=conf/train_lm.yaml
+inference_asr_model=valid.acc.ave.pth
+# speed perturbation related
+# (train_set will be "${train_set}_sp" if speed_perturb_factors is specified)
+speed_perturb_factors="0.9 1.0 1.1"
+
+# NOTE: The default settings require 4 GPUs with 32 GB memory
+./asr_streaming.sh \
```
Comment (sw005320) [inline egs2/laborotv/asr1/run_streaming.sh:21]: ditto

File: espnet/nets/batch_beam_search_online.py [hunk 7]
```diff
@@ -0,0 +1,289 @@
+"""Parallel beam search module for online simulation."""
```
Comment (sw005320) [inline espnet/nets/batch_beam_search_online.py:1]: @eml914, could you review it?

File: espnet/nets/pytorch_backend/transformer/contextual_block_encoder_layer.py [hunk 8]
```diff
@@ -58,7 +58,13 @@ def __init__(
         if self.concat_after:
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/transformer/contextual_block_encoder_layer.py:58]: Please add a test to cover this change.

File: espnet/nets/pytorch_backend/transformer/embedding.py [hunk 9]
```diff
@@ -79,7 +79,30 @@ def extend_pe(self, x):
         pe = pe.unsqueeze(0)
         self.pe = pe.to(device=x.device, dtype=x.dtype)
 
-    def forward(self, x: torch.Tensor):
+    def extend_pe2(self, length, device, dtype):
```
Comment (sw005320) [inline espnet/nets/pytorch_backend/transformer/embedding.py:82]: Please add a test to cover this method.

File: espnet2/bin/asr_inference_streaming.py [hunk 10]
```diff
@@ -0,0 +1,558 @@
+#!/usr/bin/env python3
```
Comment (sw005320) [inline espnet2/bin/asr_inference_streaming.py:1]: I think this inference design may need some discussion.
It would be great if @hfujihara, @eml914, @b-flo, @kamo-naoyuki share your ideas given your use case.

File: espnet2/asr/encoder/contextual_block_transformer_encoder.py [hunk 11]
```diff
@@ -179,6 +179,28 @@ def forward(
         xs_pad: torch.Tensor,
         ilens: torch.Tensor,
         prev_states: torch.Tensor = None,
+        is_final = True
```
Comment (sw005320) [inline espnet2/asr/encoder/contextual_block_transformer_encoder.py:182]: @eml914, can you review it?

File: espnet2/asr/encoder/contextual_block_transformer_encoder.py [hunk 12]
```diff
@@ -325,3 +347,342 @@ def forward(
...
+                "buffer_after_downsampling": buffer_after_downsampling,
+                "masks_buffer": masks_buffer,
+                "n_processed_blocks": n_processed_blocks+block_num,
+                "past_encoder_ctx": past_encoder_ctx
+            }
+        
+        return ys_pad, olens, next_states
+
+    # Refactored implementation of Emiru's code
+    def forward_train2(
```
Comment (sw005320) [inline espnet2/asr/encoder/contextual_block_transformer_encoder.py:577]: @hfujihara, Is this used?
Do you mean that it is better to use this?

**Review by sw005320 — COMMENTED** [review]
I had some shallow reviews.
In general, I recommend you add a test as much as possible.

**Comment by sw005320** [comment]
Can we also use this as a base for streaming NA?
@YosukeHiguchi, @Jzmo, could you give us some comments?

**Comment by sw005320** [comment]
@hfujihara and @D-Keqi,
Can you also upload the screenshot video of this streaming?
I think this is quite impressive.

File: egs2/aishell/asr1/conf/decode_asr_streaming2.yaml (hunk 2 above)
Comment (D-Keqi) [reply egs2/aishell/asr1/conf/decode_asr_streaming2.yaml:1]: the streaming2 is for the previous streaming method (not real streaming). If there is no need to keep it, I can delete it.

File: egs2/aishell/asr1/conf/train_asr_streaming_conformer.yaml (hunk 3 above)
Comment (D-Keqi) [reply egs2/aishell/asr1/conf/train_asr_streaming_conformer.yaml:1]: Sure!

**Comment by D-Keqi** [comment]
Sorry, there are some issues in the previous PR. Please refer to the one uploaded this time.

File: espnet/nets/batch_beam_search_online.py [hunk 13]
```diff
@@ -0,0 +1,279 @@
...
+        
+    def extend(self, x: torch.Tensor, hyps: Hypothesis) -> List[Hypothesis]:
+        """Extend probabilities and states with more encoded chunks.
+
+        Args:
+            x (torch.Tensor): The extended encoder output feature
+            hyps (Hypothesis): Current list of hypothesis
+
+        Returns:
+            Hypothesis: The exxtended hypothesis
```
Comment (ShigekiKarita) [inline espnet/nets/batch_beam_search_online.py:272]: typo: extended.

btw can you briefly summarize what kind of methods are required (e.g., extend_prob, extend_state) in the class-level (or module-level) docstring?

File: egs2/aishell/asr1/README.md [hunk 14]
```diff
@@ -1,3 +1,44 @@
+# Streaming Conformer + specaug + speed perturbation: feats=raw, n_fft=512, hop_length=128
```
Comment (pzelasko) [inline egs2/aishell/asr1/README.md:1]: I'm curious if the real-time factor metric is also available for these models (probably for a specific GPU / CPU).

**Comment by sw005320** [comment]
I recommend you summarize the TODO items with the check box in the initial comment field in this PR, e.g., 
https://github.com/espnet/espnet/pull/3613

File: espnet2/asr/encoder/contextual_block_transformer_encoder.py (hunk 12 above)
Comment (hfujihara) [reply espnet2/asr/encoder/contextual_block_transformer_encoder.py]: This is not used and I think we could delete this for this PR.
Maybe in the future, it would be better if we could replace this with forward_train (Emiru's original implementation), but to do this it needs to be tested more to make sure it works exactly the same way in training.

**Comment by hfujihara** [comment]
> @hfujihara and @D-Keqi,
> Can you also upload the screenshot video of this streaming?
> I think this is quite impressive.

@sw005320 
You mean this one?
https://user-images.githubusercontent.com/16097442/134440861-4d0712f9-119f-4c29-80bc-f72c28938bcd.mov

File: espnet2/asr/encoder/contextual_block_transformer_encoder.py (hunk 12 above)
Comment (D-Keqi) [reply espnet2/asr/encoder/contextual_block_transformer_encoder.py]: Yes, I have deleted this in the resubmitted PR. I'm sorry I make some mistakes when submitting the first PR.

File: egs2/TEMPLATE/asr1/asr_streaming.sh (hunk 1 above)
Comment (eml914) [reply egs2/TEMPLATE/asr1/asr_streaming.sh:1]: I agree with @sw005320 because it is harder to maintain both the `asr.sh` and `asr_streaming.sh`.  The training part can use `asr.sh`.

**Comment by D-Keqi** [comment]
> @D-Keqi, I recommend you summarize the TODO items with the check box in the initial comment field in this PR, e.g.,
> #3613

OK!

**Comment by sw005320** [comment]
Can you fix the CI error?

**Comment by D-Keqi** [comment]
> Can you fix the CI error?

Sure!

**Comment by Emrys365** [comment]
Just a minor comment:

It seems that the normalization part is still following the offline style, e.g. UtteranceMVN and GlobalMVN.
Will you consider adding the online normalization in the future? (Reference: [Online normalization for training neural networks](https://proceedings.neurips.cc/paper/2019/file/cb3ce9b06932da6faaa7fc70d5b5d2f4-Paper.pdf))
> Open-source code can also be found: https://github.com/Cerebras/online-normalization

**Comment by Daniel-asr** [comment]
Hi everyone and thank you for your great efforts to implement streaming for ESPnet.

I've been trying to test the new streaming capabilities (printing text while listening to a live recording) but could not find how to. (I tried your [Google Colab](https://colab.research.google.com/github/espnet/notebook/blob/master/espnet2_asr_realtime_demo.ipynb#scrollTo=nRo9Gx__BT1-) but it's not in the demo yet)

I found that you published two streaming models on the readme [ASR demo](https://github.com/espnet/espnet#asr-demo) for the tedlium2 dataset but I can't test it because I couldn't find recog.wav in the tedlium folder.

I would appreciate a direction for testing the streaming feature.

My final goal is to try it with a model I trained, so my first question is: If I want to implement streaming for a model I train does it have to be trained on asr_streaming.sh?

Thank you in advance!

**Comment by sw005320** [comment]
Sorry for our missing information.

We actually finished the first version in https://github.com/espnet/notebook/blob/master/espnet2_streaming_asr_demo.ipynb
This is only Chinese, but the tedlium2 may work.
We're not working on 
- [ ] adding more documents
- [ ] extending more models than aishell Mandarin models in the notebook
- [ ] fix some known issues (Combine asr.sh and asr_streaming.sh, WER evaluation, Layer normalization)

If you have any suggestions, that would be great.

**Comment by Emrys365** [comment]
> Just a minor comment:
> 
> It seems that the normalization part is still following the offline style, e.g. UtteranceMVN and GlobalMVN. Will you consider adding the online normalization in the future? (Reference: [Online normalization for training neural networks](https://proceedings.neurips.cc/paper/2019/file/cb3ce9b06932da6faaa7fc70d5b5d2f4-Paper.pdf))
> 
> > Open-source code can also be found: https://github.com/Cerebras/online-normalization

> Just an update :)
The implementation in Asteroid is also a good reference: https://github.com/asteroid-team/asteroid-filterbanks/blob/master/asteroid_filterbanks/pcen.py#L12
